spec.md

```diff
@@ -0,0 +1,123 @@
+# SBD Spec
+
+This SBD spec defines the protocol for SBD servers and clients to communicate.
+SBD is a simple websocket-based message relay protocol. 
+
+## 1. Websocket Stack
+
+The SBD protocol is built upon websockets.
+
+### 1.1. Websocket Configuration
+
+#### 1.1.1. Message and Frame Size
+
+The maximum SBD message size (including 32 byte header) is 20000 bytes.
+
+SBD clients and servers MAY set the max message size in the websocket library to 20000 to help enforce this.
+
+The maximum frame size MUST be set larger than 20000 so that sbd messages always fit in a single websocket frame.
+
+## 2. Cryptography
+
+### 2.1. TLS
+
+Generally over the WAN SBD servers SHOULD be available over TLS (wss://), and on a LAN without (ws://).
+
+### 2.2. Ed25519
+
+Clients will be identified by ed25519 public key. Client sessions will be validated by ed25519 signature.
+
+## 3. Protocol
+
+### 3.1. Connect Path
+
+Clients MUST specify exactly 1 http path item on the websocket connection url.
+This item must be the base64url encoded public key that this client will be identified by.
+This public key MUST be unique to this new connection.
+
+### 3.2. Messages
+
+#### 3.2.1. Header
+
+SBD messages always contain first a 32 byte header. Messages less than 32 bytes are invalid.
+
+If the header starts with 28 zero bytes, the message is a "command". Otherwise, the message is a "forward".
+
+If the header is a "command" the next four literal ascii bytes are interpreted as the command type:
+
+- `lbrt` - limit byte nanos - 4 byte i32be limit - server sent
+- `lidl` - limit idle millis - 4 byte i32be limit - server sent
+- `areq` - authentication request - 32 byte nonce - server sent
+- `ares` - authentication response - 64 byte signature - client sent
+- `srdy` - server ready - no additional data - server sent
+- `keep` - keepalive - no additional data - client sent
+
+If the header is a "forward" type, the 32 byte header is interpreted as the public key to forward the message to.
+The remaining bytes in the message are the data to forward.
+
+#### 3.2.2. Forward
+
+When a client sends a "forward" message to the server, the first 32 bytes represent the
+peer the message should be forwarded to.
+
+When a server sends a "forward" message to the client the message should be forwarded to
+the first 32 bytes will be altered to represent the peer from which the message originated.
+
+#### 3.2.3. Flow
```

Comment:
Where does connection limiting happen in this flow? I assume it's right before, so if the server accepts the network connection then it has capacity before its connection limit to handle that connection

Reply:
I added a line as the first stage in the flow about optionally dropping connections at the tcp stage. Let me know if that makes sense/was what you were looking for.